src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -554,32 +563,57 @@ private Task ProcessHeadersFrameAsync<TContext>(IHttpApplication<TContext> appli
                 }
 
                 // Start a new stream
-                _currentHeadersStream = new Http2Stream<TContext>(application, new Http2StreamContext
-                {
-                    ConnectionId = ConnectionId,
-                    StreamId = _incomingFrame.StreamId,
-                    ServiceContext = _context.ServiceContext,
-                    ConnectionFeatures = _context.ConnectionFeatures,
-                    MemoryPool = _context.MemoryPool,
-                    LocalEndPoint = _context.LocalEndPoint,
-                    RemoteEndPoint = _context.RemoteEndPoint,
-                    StreamLifetimeHandler = this,
-                    ClientPeerSettings = _clientSettings,
-                    ServerPeerSettings = _serverSettings,
-                    FrameWriter = _frameWriter,
-                    ConnectionInputFlowControl = _inputFlowControl,
-                    ConnectionOutputFlowControl = _outputFlowControl,
-                    TimeoutControl = TimeoutControl,
-                });
-
-                _currentHeadersStream.Reset();
+                _currentHeadersStream = GetStream(application);
+
                 _headerFlags = _incomingFrame.HeadersFlags;
 
                 var headersPayload = payload.Slice(0, _incomingFrame.HeadersPayloadLength); // Minus padding
                 return DecodeHeadersAsync(_incomingFrame.HeadersEndHeaders, headersPayload);
             }
         }
 
+        private Http2Stream GetStream<TContext>(IHttpApplication<TContext> application)
+        {
+            if (_streamPool.TryPop(out var stream))
+            {
+                stream.Initialize(CreateHttp2StreamContext());
+                return stream;
+            }
+
+            return new Http2Stream<TContext>(
+                    application,
+                    CreateHttp2StreamContext());
+        }
+
+        private Http2StreamContext CreateHttp2StreamContext()
+        {
+            return new Http2StreamContext
+            {
+                ConnectionId = ConnectionId,
+                StreamId = _incomingFrame.StreamId,
+                ServiceContext = _context.ServiceContext,
+                ConnectionFeatures = _context.ConnectionFeatures,
+                MemoryPool = _context.MemoryPool,
+                LocalEndPoint = _context.LocalEndPoint,
+                RemoteEndPoint = _context.RemoteEndPoint,
+                StreamLifetimeHandler = this,
+                ClientPeerSettings = _clientSettings,
+                ServerPeerSettings = _serverSettings,
+                FrameWriter = _frameWriter,
+                ConnectionInputFlowControl = _inputFlowControl,
+                ConnectionOutputFlowControl = _outputFlowControl,
+                TimeoutControl = TimeoutControl,
+            };
+        }
+
+        private void ReturnStream(Http2Stream stream)
+        {
+            if (_streamPool.Count < MaxStreamPoolSize)
```

Comment:
Why isn't MaxStreamPoolSize managed inside Push?

Reply:
When I call Push(), I expect it to always push something, rather than sometimes. Renaming to TryPush and passing in the MaxStreamCount to the stack doesn't feel great either.

I personally prefer this as it more closely resembles an actual stack and follows the usage patterns. Deciding whether it should or shouldn't be pushed is a responsibility outside of Push(). If you have strong pushback I can change it.

Reply:
It's ok so long as this check only ever happens in one place.